ledger/streaming/client/src/ledgerdb_sink.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2018-2021 The MobileCoin Foundation
+#![allow(dead_code)]
+
+use futures::stream::{Stream, StreamExt};
+use mc_common::logger::{log, Logger};
+use mc_ledger_db::Ledger;
+use mc_ledger_streaming_api::StreamResult;
+use mc_transaction_core::BlockData;
+use pin_project::pin_project;
+use std::sync::{Arc, Mutex};
+use tokio::sync::mpsc;
+/// A block sink that takes blocks from a passed stream and puts them into
+/// ledger db. This sink should live downstream from a verification source that
+/// has already done block content, scp, and avr validation and thus is
+/// considered a trusted stream
+#[pin_project]
+pub struct BlockSink<BS: Stream<Item = StreamResult<BlockData>> + std::marker::Unpin> {
+    /// Stream of block data
+    stream: BS,
+
+    /// Message channel that sends data to the db
+    sender: mpsc::Sender<BlockData>,
+
+    /// Logger running
+    logger: Logger,
+}
+
+impl<BS: Stream<Item = StreamResult<BlockData>> + std::marker::Unpin> BlockSink<BS> {
+    pub fn new(stream: BS, logger: Logger, sender: mpsc::Sender<BlockData>) -> Self {
+        Self {
+            stream,
+            logger,
+            sender,
+        }
+    }
+
+    // Ingest blocks from stream into a sink for processing
+    async fn broadcast(&mut self) {
+        while let Some(block) = self.stream.next().await {
+            if let Ok(block) = block {
+                let index = block.block().index;
+                match self.sender.send(block).await {
+                    Ok(_) => log::debug!(self.logger, "successfully sent block {}", index),
+                    Err(_) => log::warn!(self.logger, "receiver failed to send"),
+                }
+            }
+        }
+    }
+}
+
+/// Write to the ledger in a synchronous loop that consistently polls for blocks
+pub fn sink(
+    ledger_ref: Arc<Mutex<impl Ledger>>,
+    mut receiver: mpsc::Receiver<BlockData>,
+    logger: Logger,
+) {
+    let mut ledger = ledger_ref.lock().unwrap();
+    loop {
+        if let Some(block_data) = receiver.blocking_recv() {
+            let index = block_data.block().index;
+            ledger
+                .append_block(
+                    block_data.block(),
+                    block_data.contents(),
+                    block_data.signature().clone(),
+                )
+                .unwrap();
+            log::trace!(logger, "Block {} written", index)
+        } else {
+            log::warn!(logger, "Receiver shut down!");
```

Comment:
Is that the only reason we might get an error? It seems better to capture and log the error here.

Reply:
Yeah this definitely a todo

Reply:
Then let's add a TODO ;)